src/vsr/replica.zig

```diff
@@ -370,33 +350,62 @@ pub fn ReplicaType(
             while (!self.opened) self.superblock.storage.tick();
 
-            if (self.journal.is_empty()) {
-                // The data file is brand new — no messages have ever been written.
-                // Transition to normal status; no need to run the VSR recovery protocol.
-                assert(self.journal.dirty.count == 0);
-                assert(self.journal.faulty.count == 0);
-                assert(self.commit_min == 0);
-                assert(self.commit_max == 0);
-                assert(self.op_checkpoint == 0);
-                assert(self.op == 0);
-                assert(self.view == 0);
-
-                log.debug("{}: open: empty data file", .{self.replica});
-                self.transition_to_normal_from_recovering_status(0);
-                assert(self.status == .normal);
-            } else if (self.replica_count == 1) {
-                if (self.journal.faulty.count != 0) @panic("journal is corrupt");
+            const vsr_headers = self.superblock.working.vsr_headers();
+            var op_head: u64 = vsr_headers.slice[0].op;
+            for (self.journal.headers) |*header| {
+                if (header.command == .prepare and header.op > op_head) {
+                    assert(header.view <= self.log_view);
+                    op_head = header.op;
+                }
+            }
+
+            self.op = op_head;
+            for (vsr_headers.slice) |*header| self.replace_header(header);
+
+            if (self.op == 0 and self.op_checkpoint == 0 and
+                self.journal.header_with_op(self.op) == null)
+            {
+                // Empty log, but slot 0 is corrupt.
+                // replace_header() didn't repair it because op 0 is "committed".
+                const root_prepare = Header.root_prepare(self.superblock.working.cluster);
+                self.journal.set_header_as_dirty(&root_prepare);
+            }
+
+            const header_head = self.journal.header_with_op(self.op).?;
+            assert(header_head.view <= self.superblock.working.vsr_state.log_view);
+
+            if (self.replica_count == 1) {
+                if (self.journal.faulty.count > 0) {
+                    @panic("journal is corrupt");
+                }
+                assert(self.op_head_certain());
+
+                if (self.commit_min < self.op) {
+                    self.commit_journal(self.op);
+                } else {
+                    self.transition_to_normal_from_recovering_status();
+                }
             } else {
-                assert(self.status == .recovering);
+                if (self.op_head_certain()) {
+                    if (self.view == self.log_view) {
+                        if (self.primary_index(self.view) == self.replica) {
+                            self.transition_to_view_change_status(self.view + 1);
+                        } else {
+                            self.transition_to_normal_from_recovering_status();
```

Comment:
I'm curious why there's no `if (self.commit_min < self.op)` in here.

Reply:
I'm not sure what you mean exactly; can you elaborate?

Reply:
A cluster-of-1 is effectively the primary, which must be caught up on commits before it can start preparing more ops.

But in this case, the replica is a follower — it can commit asynchronously while also preparing/acking ops for the primary.